### ISSUE TITLE
More Sacado build errors

### ISSUE THREAD

**Issue by ibaned** [kokkos/kokkos#934]
The following are from the `serial` build on `shepard`

```
/home/daibane/serial/trilinos-update/packages/sacado/src/Kokkos_ViewFactory.hpp(105): error: no instance of overloaded function "Kokkos::dimension_scalar" matches the argument list
            argument types are: (const Kokkos::DynRankView<double, Kokkos::DefaultExecutionSpace>, const Kokkos::DynRankView<Sacado::Fad::DFad<double>, Kokkos::DefaultExecutionSpace>)
      layout.dimension[rank] = dimension_scalar(views...);
                               ^
          detected during:
            instantiation of "ResultView Kokkos::ViewFactory<ViewPack...>::create_view<ResultView,CtorProp,Dims...>(const ViewPack &..., const CtorProp &, Dims...) [with ViewPack=<Kokkos::DynRankView<double, Kokkos::DefaultExecutionSpace>, Kokkos::DynRankView<Sacado::Fad::DFad<double>, Kokkos::DefaultExecutionSpace>>, ResultView=Kokkos::DynRankView<Sacado::Fad::DFad<double>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>>, CtorProp=char [4],
                      Dims=<Kokkos::HostSpace::size_type={size_t={unsigned long}}>]" at line 100 of "/home/daibane/serial/trilinos-update/packages/sacado/example/view_factory_example.cpp"
            instantiation of "Kokkos::DynRankView<Kokkos::ViewFactory<View1, View2>::value_type, View1::device_type> my_func_dynamic(const View1 &, const View2 &) [with View1=Kokkos::DynRankView<double, Kokkos::DefaultExecutionSpace>, View2=Kokkos::DynRankView<Sacado::Fad::DFad<double>, Kokkos::DefaultExecutionSpace>]" at line 181 of "/home/daibane/serial/trilinos-update/packages/sacado/example/view_factory_example.cpp"

compilation aborted for /home/daibane/serial/trilinos-update/packages/sacado/example/view_factory_example.cpp (code 2)
```

```
In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(104): error: namespace "Kokkos::Experimental::Impl" has no member "ViewDataType"
    typedef typename Kokkos::Experimental::Impl::ViewDataType< flat_value_type , dimension >::type flat_data_type;
```

```
In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(199): error #3202: class "Kokkos::Impl::ViewDataAnalysis<DataType, ArrayLayout, Sacado::MP::Vector<StorageType>>" cannot be partially specialized in the current scope
  struct ViewDataAnalysis< DataType     /* Original view data type */
         ^

```

```
In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(318): error: ViewValueFunctor is not a template
      typedef ViewValueFunctor< ExecSpace, scalar_type > FunctorType ;
              ^
```

```
In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(515): error #3202: class "Kokkos::Impl::ViewMapping<Traits, std::enable_if<<expression>, void>::type>" cannot be partially specialized in the current scope
  class ViewMapping< Traits , /* View internal mapping */
        ^
```

```
In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(545): error: qualified name is not allowed
    typedef Impl::MPVectorAllocation<sacado_mp_vector_type> handle_type;
            ^
```

```
n file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(893): error #3202: class "Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>" cannot be partially specialized in the current scope
  class ViewMapping< DstTraits , SrcTraits ,
        ^

In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(974): error #3202: class "Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>" cannot be partially specialized in the current scope
  class ViewMapping< DstTraits , SrcTraits ,
        ^
```

```
In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(1197): error #3202: class "Kokkos::Impl::ViewMapping<std::enable_if<<expression>, void>::type, Kokkos::ViewTraits<DataType, P...>, Arg0, Args...>" cannot be partially specialized in the current scope
  struct ViewMapping
         ^

In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(1341): error #3202: class "Kokkos::Impl::ViewMapping<void, Kokkos::ViewTraits<DataType, P...>, Sacado::MP::VectorPartition<Size>>" cannot be partially specialized in the current scope
  class ViewMapping<
        ^

In file included from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Kokkos_View_MP_Vector.hpp(53),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/Stokhos_Sacado_Kokkos_MP_Vector.hpp(62),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp(49),
                 from /home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_ETI_Helpers_MP_Vector.hpp(43),
                 from /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp(45):
/home/daibane/serial/trilinos-update/packages/stokhos/src/sacado/kokkos/vector/KokkosExp_View_MP_Vector_Contiguous.hpp(1513): error #3202: class "Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>" cannot be partially specialized in the current scope
  class ViewMapping< DstTraits , SrcTraits ,
        ^
```

```
compilation aborted for /home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Vector_MP_Vector_Serial.cpp (code 4)
```

```
/home/daibane/serial/trilinos-update/packages/kokkos/core/src/Kokkos_View.hpp(498): error: class "Kokkos::Impl::ViewMapping<Kokkos::ViewTraits<Sacado::MP::Vector<Tpetra::SFS_int_double_4_DFN_CPU_int_int_Kokkos_Compat_KokkosSerialWrapperNode> **, Kokkos::LayoutLeft, Kokkos::Serial::execution_space, void>, void>" has no member "Rank"
    enum { Rank = map_type::Rank };
                            ^
          detected during:
            instantiation of class "Kokkos::View<DataType, Properties...> [with DataType=Sacado::MP::Vector<Tpetra::SFS_int_double_4_DFN_CPU_int_int_Kokkos_Compat_KokkosSerialWrapperNode> **, Properties=<Kokkos::LayoutLeft, Kokkos::Serial::execution_space, void>]" at line 115 of "/home/daibane/serial/trilinos-update/packages/kokkos/containers/src/Kokkos_DualView.hpp"
            instantiation of class "Kokkos::DualView<DataType, Arg1Type, Arg2Type, Arg3Type> [with DataType=Sacado::MP::Vector<Tpetra::SFS_int_double_4_DFN_CPU_int_int_Kokkos_Compat_KokkosSerialWrapperNode> **, Arg1Type=Kokkos::LayoutLeft, Arg2Type=Kokkos::Serial::execution_space, Arg3Type=void]" at line 607 of "/home/daibane/serial/trilinos-update/packages/tpetra/core/src/Tpetra_MultiVector_decl.hpp"
            instantiation of class "Tpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node, classic> [with Scalar=Sacado::MP::Vector<Tpetra::SFS_int_double_4_DFN_CPU_int_int_Kokkos_Compat_KokkosSerialWrapperNode>, LocalOrdinal=int, GlobalOrdinal=int, Node=Tpetra::Kokkos_Compat_KokkosSerialWrapperNode, classic=false]" at line 175 of "/home/daibane/serial/trilinos-update/packages/tpetra/core/src/Tpetra_Experimental_BlockMultiVector_decl.hpp"
            instantiation of class "Tpetra::Experimental::BlockMultiVector<Scalar, LO, GO, Node> [with Scalar=Sacado::MP::Vector<Tpetra::SFS_int_double_4_DFN_CPU_int_int_Kokkos_Compat_KokkosSerialWrapperNode>, LO=int, GO=int, Node=Tpetra::Kokkos_Compat_KokkosSerialWrapperNode]" at line 80 of "/home/daibane/serial/trilinos-update/packages/tpetra/core/src/Tpetra_Experimental_BlockVector_decl.hpp"
            instantiation of class "Tpetra::Experimental::BlockVector<Scalar, LO, GO, Node> [with Scalar=Sacado::MP::Vector<Tpetra::SFS_int_double_4_DFN_CPU_int_int_Kokkos_Compat_KokkosSerialWrapperNode>, LO=int, GO=int, Node=Tpetra::Kokkos_Compat_KokkosSerialWrapperNode]" at line 54 of "/home/daibane/serial/build_updated/packages/stokhos/src/Tpetra_Experimental_BlockVector_MP_Vector_Serial.cpp"

```

**Comment by ibaned** [comment]
I've got some of these fixed, WIP...

**Comment by ibaned** [comment]
Addressed many of these by trilinos/Trilinos@fa73121.
